test/puppetlabs/http/client/metrics_test.clj

```diff
@@ -0,0 +1,481 @@
+                    bytes-read-id-with-verb (str base-metric-id ".GET" ".bytes-read")]
+                (is (= (set (list bytes-read-id bytes-read-id-with-verb))
+                       (set (keys client-metrics))
+                       (set (keys client-metrics-data))))
+                (is (every? #(instance? Timer %) (vals client-metrics)))
+                (let [bytes-read-data (get client-metrics-data bytes-read-id)]
+                  (is (every? #(instance? ClientMetricData %) (vals client-metrics-data)))
+
+                  (is (= 1 (.getCount bytes-read-data)))
+                  (is (= bytes-read-id (.getMetricId bytes-read-data)))
+                  (is (<= 1000 (.getMean bytes-read-data)))
+                  (is (<= 1000 (.getAggregate bytes-read-data))))))))))
+     (testing "metrics work for failed request"
+       (try
+         (testwebserver/with-test-webserver-and-config
+          (unbuffered-test/blocking-handler data) port {:shutdown-timeout-seconds 1}
+          (let [metric-registry (MetricRegistry.)]
+            (with-open [client (-> (ClientOptions.)
+                                   (.setSocketTimeoutMilliseconds 200)
+                                   (.setConnectTimeoutMilliseconds 100)
+                                   (Async/createClient metric-registry))]
+              (let [request-options (doto (RequestOptions. (str "http://localhost:" port "/hello"))
+                                      (.setAs ResponseBodyType/UNBUFFERED_STREAM))
+                    response (-> client (.get request-options) .deref)
+                    error (.getError response)
+                    body (.getBody response)]
+                (is (nil? error))
+                (is (thrown? SocketTimeoutException (slurp body)))
+                (let [client-metrics (.getClientMetrics client)
+                      client-metrics-data (.getClientMetricsData client)
+                      base-metric-id (str "puppetlabs.http-client.experimental.with-url.http://localhost:" port "/hello")
+                      bytes-read-id (str base-metric-id ".bytes-read")
+                      bytes-read-id-with-verb (str base-metric-id ".GET" ".bytes-read")]
+                  (is (= (set (list bytes-read-id bytes-read-id-with-verb))
+                         (set (keys client-metrics))
+                         (set (keys client-metrics-data))))
+                  (is (every? #(instance? Timer %) (vals client-metrics)))
+                  (let [bytes-read-data (get client-metrics-data bytes-read-id)]
+                    (is (every? #(instance? ClientMetricData %) (vals client-metrics-data)))
+
+                    (is (= 1 (.getCount bytes-read-data)))
+                    (is (= bytes-read-id (.getMetricId bytes-read-data)))
+                    (is (<= 1 (.getMean bytes-read-data)))
+                    (is (<= 1 (.getAggregate bytes-read-data)))))))))
+         (catch TimeoutException e
```

Comment:
Cool, that makes sense.  My biggest concern had been about the `TimeoutException` potentially being thrown in some other cases, causing the bulk of the test to not be executed.

It does seem like there would be a way to avoid Jetty encountering an error during shutdown.  Like maybe we could inject a promise into the `blocking-handler` and have its `future` block on that promise after doing the stream write.  Once we get past having made all of the assertions from within this test, could deliver on that promise.  The `future` could then be unblocked and close its output stream.  If all this happens before the webserver is shut down, I'd hope the server could shut down cleanly - without timing out - and that we wouldn't have to artificially try to work around that anymore be setting `shutdown-timeout-seconds` to 1 in the web server configuration.

That said, I think you're just following the pattern of pre-existing unbuffered stream tests here and would be fine with deferring any broader cleanup around that set of tests to sometime later than this PR.